test/appStartup.test.ts

```diff
@@ -0,0 +1,160 @@
+        coreApi = await import('@azure/functions-core');
+    });
+
+    after(() => {
+        process.env = originalEnv;
+    });
+
+    afterEach(async () => {
+        await stream.afterEachEventHandlerTest();
+        coreApi.Disposable.from(...testDisposables).dispose();
+        testDisposables = [];
+        process.chdir(originalCwd);
+    });
+
+    it('runs app startup hooks in non-specialization scenario', async () => {
+        const hostVersion = '2.7.0';
+        const functionAppDirectory = __dirname;
+        const expectedStartupContext: coreTypes.AppStartupContext = {
+            functionAppDirectory,
+            hostVersion,
+            hookData: {},
+        };
+
+        const startupFunc = sinon.spy();
+        testDisposables.push(coreApi.registerHook('appStartup', startupFunc));
+
+        stream.addTestMessage(WorkerInitMsg.init(functionAppDirectory, hostVersion));
+
+        await stream.assertCalledWith(
+            WorkerInitMsg.receivedInitLog,
+            WorkerInitMsg.warning('Worker failed to load package.json: file does not exist'),
+            Msg.executingHooksLog(1, 'appStartup'),
+            Msg.executedHooksLog('appStartup'),
+            WorkerInitMsg.response
+        );
+
+        expect(startupFunc.callCount).to.be.equal(1);
+        expect(startupFunc.args[0][0]).to.deep.equal(expectedStartupContext);
+    });
+
+    it('runs app startup hooks only once in specialiation scenario', async () => {
+        const hostVersion = '2.7.0';
+        const functionAppDirectory = __dirname;
+        const expectedStartupContext: coreTypes.AppStartupContext = {
+            functionAppDirectory,
+            hostVersion,
+            hookData: {},
+        };
+        const startupFunc = sinon.spy();
+
+        stream.addTestMessage(WorkerInitMsg.init(functionAppDirectory, hostVersion));
+        await stream.assertCalledWith(
+            WorkerInitMsg.receivedInitLog,
+            WorkerInitMsg.warning('Worker failed to load package.json: file does not exist'),
+            WorkerInitMsg.response
+        );
+
+        testDisposables.push(coreApi.registerHook('appStartup', startupFunc));
+
+        stream.addTestMessage({
+            requestId: 'id',
+            functionEnvironmentReloadRequest: {
+                functionAppDirectory,
+            },
+        });
+        await stream.assertCalledWith(
+            EnvReloadMsg.reloadEnvVarsLog(0),
+            EnvReloadMsg.changingCwdLog(functionAppDirectory),
+            WorkerInitMsg.warning('Worker failed to load package.json: file does not exist'),
+            Msg.executingHooksLog(1, 'appStartup'),
+            Msg.executedHooksLog('appStartup'),
+            EnvReloadMsg.reloadSuccess
+        );
+
+        expect(startupFunc.callCount).to.be.equal(1);
+        expect(startupFunc.args[0][0]).to.deep.equal(expectedStartupContext);
+    });
+
+    it('persists hookData changes from app startup hooks in worker channel', async () => {
```

Comment:
worker channel is an internal detail, and there's no reason we have to persist hook data on it long term. If you want to include this test it should check that appHookData is persisted for an invocation hook, which we do need to support long term

Reply:
My purpose of this test was to ensure that the app-level `hookData` was persisted in the app scope. A better test would have been to ensure that a hypothetical app teardown hook sees changes that app start hooks make, but that hypothetical hook doesn't exist yet. I rewrote the title of this test though. Does it make more sense now?